### ISSUE TITLE
Option to get a current state of draft as plain object

### ISSUE THREAD

**Issue by user753** [immerjs/immer#441]
## 🚀 Feature Proposal

Function to get a current state of draft as plain object

## Motivation

I would like to pass the current state of draft and don't worry if it would be modified.

## Example

```
produce({value: 1}, draft => {
  draft.value = 2
  const state = toJS(draft) // would be {value:2}
  draft.value = 3
})
```

**Comment by mweststrate** [comment]
Repeating the proposal as motivation doesn't count ;-)

I guess a simple utility like lodash.deepClone or
JSON.parse(JSON.stringify(draft)) can do the trick already. Due to internal
administration and optimizations, our implementation doesn't really allow
for drafts to finilizad, with structural sharing, multiple times.

**Comment by user753** [comment]
@mweststrate Does it mean it isn't possible to do something better than `lodash.deepClone`?
For example if we could check if subtree of draft `isModified`, we could do something like 
```
  const state = {}
  for(const key in draft){
    state[key] = isModified(draft[key]) ? deepClone(draft[key]) : original(draft[key])
  }
```

**Comment by mweststrate** [comment]
Conceptually it is easy, I think in practice it will be quite tricky. But
maybe I'm wrong and it is quite simple. Anyway, attempts are welcome :)

**Comment by illuminist** [comment]
As for my usage, I had a memoized helper function being used everywhere in my code. Now, after I need to alter a state I need to call that helper function to group item to determine how to alter the second state. The problem is that memoized helper is being called first time with newly alter state so it memoizes grouped items as draft objects. When the helper is called somewhere else after that, the helper will return draft objects which causes an error. I currently use `lodash.deepClone` to input newly cloned state to that helper as a workaround.

**Comment by lostpebble** [comment]
For my use-case, its about composability:

For example:

```ts
const store = {
  creatingItem: {
    id: null,
    name: "James",
    points: 2,
  },
  finishedItems: [],
  totalPoints: 0,
};

const calculateTotalPoints = (s, o) => {
  let total = 0;

  for (const item of o.finishedItems) {
    total += item.points;
  }

  s.totalPoints = total;
}

const finalizeItem = (s, o) => {
  const newItem = {
    ...o.creatingItem,
    id: generateId(),
  };
  s.finishedItems.push(newItem);

  calculateTotalPoints(s, o); // <- "o" here is representing the stale state
}

const deleteItem = (itemIndex) => (s, o) => {
  s.finishedItems.splice(itemIndex, 1);
  calculateTotalPoints(s, o); // <- "o" here is representing the stale state
}
```

Those functions are "state updaters" for my state library, they're basically immer producers but with the current original state passed in as the second argument (for referencing, for performance reasons).

Its a bit of a contrived example, but I run into this kind of thing with composabilty quite a bit. I end up having to do things like:

```ts
Store.update([finalizeItem, calculateTotalPoints])
```

```ts
Store.update([deleteItem, calculateTotalPoints])
```

Basically, composing these updates as a sequence - but that means I need to remember to compose them like that in each interface interaction. Would be _much_ nicer if I could compose the changes from directly inside the "updaters" as they are needed. For this, I need to be able to create a fresh "original" state from the current draft state.

**Comment by mweststrate** [comment]
drafts can be finalized only once. I recommend to look into createDraft and
finishDraft utilities, and skip on produce if you want to build an
abstraction that composes differently than `produce`

**Comment by lostpebble** [comment]
Okay cool thanks, will take a look.

So inside produce there's currently no way to get the current non-proxied (pure JS object) state of a draft without finalizing.

**Comment by Kaijun** [comment]
> Immer 7 has been releases and includes a `current` utility.

Hi @mweststrate  Can i only use `current` inside produce with draft object?

what's the best practice if i want to get the plain object of `proxied` variables? 

should i still use cloneDeep?

**Comment by mweststrate** [comment]
@Kaijun this issue is closed, please open a separate issue and include the problem you are trying to solve. You shouldn't have any proxied variables _outside_ produce.
